validator/src/journal/block_manager.rs

```diff
@@ -0,0 +1,303 @@
+                Box::new(iter)
+            }
+            None => Box::new(::std::iter::empty()),
+        }
+    }
+
+    fn contains(&self, block_id: &str) -> bool {
+        self.anchors_by_block_id.contains_key(block_id)
+    }
+
+    fn increase_internal_ref_count(&mut self, block_id: &str) {
+        if let Some(anchor) = self.anchors_by_block_id.get_mut(block_id) {
+            anchor.internal_ref_count += 1;
+        }
+    }
+
+    fn decrease_internal_ref_count(&mut self, block_id: &str) {
+        if let Some(anchor) = self.anchors_by_block_id.get_mut(block_id) {
+            anchor.internal_ref_count -= 1;
+        }
+    }
+
+    fn increase_external_ref_count(&mut self, block_id: &str) {
+        if let Some(anchor) = self.anchors_by_block_id.get_mut(block_id) {
+            anchor.external_ref_count += 1;
+        }
+    }
+
+    fn decrease_external_ref_count(&mut self, block_id: &str) {
+        if let Some(anchor) = self.anchors_by_block_id.get_mut(block_id) {
+            anchor.external_ref_count -= 1;
+        }
+    }
+}
+
+struct Anchor {
+    pub blockstore_name: String,
+    pub external_ref_count: i64,
+    pub internal_ref_count: i64,
+    pub block_id: String,
+}
+
+#[derive(Default)]
+pub struct BlockManager {
+    block_by_block_id: HashMap<String, (Block, i64, i64)>,
+
+    blockstore_by_name: HashMap<String, Box<BlockStore>>,
+
+    anchors: Anchors,
+}
+
+/// The BlockManager maintains integrity of all the blocks it contains,
+/// such that for any Block within the BlockManager,
+/// that Block's predecessor is also within the BlockManager.
+impl BlockManager {
+    pub fn new() -> Self {
+        BlockManager::default()
+    }
+
+    /// Put is idempotent, making the guarantee that after put is called with a
+    /// block in the vector argument, that block is in the BlockManager
+    /// whether or not it was already in the BlockManager.
+    /// Put makes three other guarantees
+    ///     - If the zeroth block in branch does not have its predecessor
+    ///       in the BlockManager an error is returned
+    ///     - If any block after the zeroth block in branch
+    ///       does not have its predecessor as the block to its left in
+    ///       branch, an error is returned.
+    ///     - If branch is empty, an error is returned
+    pub fn put(&mut self, branch: Vec<Block>) -> Result<(), BlockManagerError> {
+        match branch.split_first() {
+            Some((head, tail)) => {
+                let predecessor_of_head_in_memory =
+                    self.block_by_block_id.contains_key(&head.previous_block_id);
+
+                let predecessor_of_head_in_any_blockstore = self.blockstore_by_name.values().any(
+                    |blockstore| blockstore.get(vec![head.previous_block_id.clone()]).count() > 0,
+                );
+                let head_is_genesis = head.previous_block_id == NULL_BLOCK_IDENTIFIER;
```

Comment:
I wouldn't use blockchain terminology here, just think of it as a tree, so this would be the root. Also, I think it should be possible to avoid doing root-node-specific checks by making the root nodes anchors.

Reply:
I'm not sure it is possible to do without the NULL_BLOCK_IDENTIFIER check. If the first put is `C 3, D 4` and with C pointing to an unknown block B is it ok or is it an error?

Reply:
I am thinking that when you create the block manager, rather than having anchors be empty, you can initialize it with an anchor to a block with id = NULL_BLOCK_IDENTIFIER